File: Top-Interview-Questions-Leetcode/C++/Easy-Collections/Single-Number.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    int singleNumber(std::vector<int>& nums) {
        // 0 ^ x = x
        // x ^ x = 0
        // x ^ y ^ z ^ x ^ z = y ^ (x ^ x) ^ (z ^ z) = y ^ 0 ^ 0 = y
        // So we can use XOR to find the single number in the array
        // where all other numbers appear twice.
        int num = 0;
        
        for (int val : nums) num ^= val;
        
        return num;
    }
};

// Example usage
int main() {
    Solution solution;
    std::vector<int> nums = {4, 1, 2, 1, 2};
    int result = solution.singleNumber(nums);
    std::cout << "The single number is: " << result << std::endl; // Output: 4
    return 0;
}

File: Top-Interview-Questions-Leetcode/Python/Easy-Collections/Single-Number.py
from typing import List

class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        num = 0
        
        for val in nums:
            num ^= val
        
        return num

# Example usage
def main():
    nums = [4, 1, 2, 1, 2]
    solution = Solution()
    result = solution.singleNumber(nums)
    print(f"The single number is: {result}")

if __name__ == "__main__":
    main()